File: 2025/03 March/22 Stickler Thief II/main.py
# Link: https://www.geeksforgeeks.org/problems/house-robber-ii/1


class Solution:
    def helper(self, nums):
        n = len(nums)
        dp = [0] * n
        dp[0] = nums[0]
        for i in range(1, n):
            pick = nums[i]
            if i > 1:
                pick += dp[i - 2]
            non = dp[i - 1]
            dp[i] = max(pick, non)
        return dp[-1]

    def maxValue(self, arr):
        n = len(arr)
        if n == 0:
            return 0
        if n == 1:
            return arr[0]
        arr1 = arr[:-1]
        arr2 = arr[1:]
        ans1 = self.helper(arr1)
        ans2 = self.helper(arr2)
        return max(ans1, ans2)


if __name__ == "__main__":
    obj = Solution()
    arr = [2, 3, 2]
    print(obj.maxValue(arr))

    arr = [1, 2, 3, 1]
    print(obj.maxValue(arr))

    arr = [2, 2, 3, 1, 2]
    print(obj.maxValue(arr))

File: 2025/03 March/22 Stickler Thief II/main.cpp
// Link: https://www.geeksforgeeks.org/problems/house-robber-ii/1

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    int getMax(vector<int> &nums, int start, int end)
    {
        int prev = 0, prev2 = 0, curr = 0;

        for (int i = start; i <= end; ++i)
        {
            curr = max(prev, prev2 + nums[i]);
            prev2 = prev;
            prev = curr;
        }

        return curr;
    }

    int maxValue(vector<int> &arr)
    {
        int n = arr.size();
        int val1 = getMax(arr, 0, n - 2);
        int val2 = getMax(arr, 1, n - 1);
        return max(val1, val2);
    }
};

int main()
{
    Solution obj;
    vector<int> arr = {2, 3, 2};
    cout << obj.maxValue(arr) << endl;

    arr = {1, 2, 3, 1};
    cout << obj.maxValue(arr) << endl;

    arr = {2, 2, 3, 1, 2};
    cout << obj.maxValue(arr) << endl;
    return 0;
}